Binman minimal implementation

UtsavAgarwalADI opened the pull request:
Introducing new kconfig options to enable binman and dynamic (compile time) u-boot environment.


image.map example:
```
ImagePos    Offset      Size  Name
00000000  00000000  009af1e6  image
00000000   00000000  00011a28  stage1-boot
00040000   00040000  0009ffc0  stage2-boot
001a0000   001a0000  0080f1e6  fitImage
 
```

Binman:

- `CONFIG_ADI_BINMAN`: enables binman, allows additional sections/configuration options to the binman node.
- `CONFIG_BINMAN_FILENAME`: file name of the resulting image file.
- `CONFIG_PACK_FITIMAGE`: enables fitImage to be packed using binman
- `CONFIG_SC5XX_FITIMAGE_NAME`: file to be used as fitImage
- `CONFIG_PACK_ROOTFS`: enables rootfs to be packed using binman
- `CONFIG_SC5XX_ROOTFS_NAME`: file to be used as rootfs

SC5XX:

- `CONFIG_SC5XX_UBOOT_SPL_OFFSET`: SPL offset, 0 default. Only affects binman image for now. 
- `CONFIG_SC5XX_UBOOT_OFFSET`: Full u-boot offset. Governs binman/full u-boot default environment.
- `CONFIG_SC5XX_FITIMAGE_OFFSET`: FitImage offset. Governs binman/full u-boot default environment.
- `CONFIG_SC5XX_ROOTFS_OFFSET`: Rootfs offset. Governs binman/full u-boot default environment.
- `CONFIG_SC5XX_LOADADDR`: Loadaddr offset. Governs binman/full u-boot default environment.

Currently, defconfigs do not enable binman by default. To enable it, `CONFIG_ADI_BINMAN `must be enabled via `menuconfig` or added to the respective defconfig.

This is done since enabling binman while compile u-boot spl gives a warning, even though the compilation succeeds.
```
Image 'image' has faked external blobs and is non-functional: u-boot.ldr

Image 'image' is missing optional external blobs but is still functional: stage2-boot

/binman/stage2-boot (u-boot.ldr):
   Missing blob

Some images are invalid
make: *** [Makefile:1133: .binman_stamp] Error 10
```

This has an image.map resulting as follows:
```
ImagePos    Offset      Size  Name
00000000  00000000  00040000  image
00000000   00000000  00011a28  stage1-boot
00040000   00040000  00000000  stage2-boot
```

The general sc5xx config options have now been added to all defconfigs as well as their respective .env files.

gmalysa commented:
Looks good to start. But I have some questions:
- Can the binman node be in a *-u-boot.dtsi and auto included? (e.g. sc598-som-ezkit-u-boot.dtsi, see https://docs.u-boot.org/en/latest/develop/devicetree/control.html#dttweaks)
- Can we replace the offsets with Kconfigs? Then we can perhaps use the same offsets in the text env to keep the two in sync by having them be symbolic and specified in the defconfig
- Does this handle missing images correctly (necessary for unit testing) or does it fail because a required image cannot be found? uboot/master is incompatible with my local python so I can't test this without doing the full six hour unit test and I'm holding off on running that for the moment
- We still need to determine the bounds between what binman does and what yocto does but we can discuss that more on Thursday

ghost commented:
>     * Can the binman node be in a *-u-boot.dtsi and auto included? (e.g. sc598-som-ezkit-u-boot.dtsi, see https://docs.u-boot.org/en/latest/develop/devicetree/control.html#dttweaks)
> 
>     * Can we replace the offsets with Kconfigs? Then we can perhaps use the same offsets in the text env to keep the two in sync by having them be symbolic and specified in the defconfig
> 

Yes, definitely. In fact, as mentioned above, I was thinking we could maybe put a basic template in ```adi-u-boot.dtsi``` and then leverage Kconfigs to adjust things per board. Ofcourse, we can always extend the node inside board specific files such as ```sc598-som-ezkit-u-boot.dtsi```

>     * Does this handle missing images correctly (necessary for unit testing) or does it fail because a required image cannot be found? uboot/master is incompatible with my local python so I can't test this without doing the full six hour unit test and I'm holding off on running that for the moment
> 

I am assuming you mean the .ldr and fitImage - unfortunately not, this is what I was rather alluding to when I mentioned that it needs all the artifacts compiled beforehand as binman needs to be the very last step.

Utsav-Agarwal committed:
[do not submit] SC5XX: Adding binman image

binman needs to be invoked after all artifacts have been compiled.

It is disabled by default.

Since the board needs ldr files to boot after processing the bootrom,
this becomes a hard requirement.

For now, all artifacts/components are listed as ext-blobs. These can be
further sorted into sections and even templates which can probably
be reused by other boads in the family.

Currently, binman is enabled by enabling CONFIG_BINMAN,
which then presents the configuration options for the final binman image.

Binman utilizes the new config options for various offsets (SPL,
u-boot full, fitImage and rootfs). This allows binman to and the
u-boot environment to be coherent.

Signed-off-by: Utsav-Agarwal_adi <utsav.agarwal@analog.com>

Utsav-Agarwal committed:
[do not submit] SC598-SOM: Adding env config for SC598-som boards

Adding new config options to the env file for dynamic compile
time values. This allows the environment to be coherent with
a binman generated image.

This commit also formats both sc598-som env files to be more
consistent.

Signed-off-by: UtsavAgarwalADI <utsav.agarwal@analog.com>

Utsav-Agarwal committed:
[do not submit] SC594-SOM: Adding env config for SC594-som boards

Adding new config options to the env file for dynamic compile
time values. This allows the environment to be coherent with
a binman generated image.

Signed-off-by: UtsavAgarwalADI <utsav.agarwal@analog.com>

Utsav-Agarwal committed:
[do not submit] SC589: Adding env config for SC589 boards

Adding new config options to the env file for dynamic compile
time values. This allows the environment to be coherent with
a binman generated image.

Signed-off-by: UtsavAgarwalADI <utsav.agarwal@analog.com>

Utsav-Agarwal committed:
[do not submit] SC584: Adding env config for SC584 boards

Adding new config options to the env file for dynamic compile
time values. This allows the environment to be coherent with
a binman generated image.

Signed-off-by: UtsavAgarwalADI <utsav.agarwal@analog.com>

Utsav-Agarwal committed:
[do not submit] SC573: Adding env config for SC573 boards

Adding new config options to the env file for dynamic compile
time values. This allows the environment to be coherent with
a binman generated image.

Signed-off-by: UtsavAgarwalADI <utsav.agarwal@analog.com>

ghost commented:
Hi Greg, 

I have updated the comment on the changes made/usage. 

I'm slightly unsure of how we want to enable binman and if I should just use CONFIG_BINMAN instead of CONFIG_ADI_BINMAN to enable our binman options - this just allows me to easily toggle it in the menu as opposed to passing CONFIG_BINMAN=y on the command line.

ghost commented:
As discussed, removed ADI_BINMAN
